src/main/java/software/amazon/cloudwatchlogs/emf/model/MetricDirective.java

```diff
@@ -60,11 +61,15 @@ void putMetric(String key, double value) {
     }
 
     void putMetric(String key, double value, Unit unit) {
-        if (metrics.containsKey(key)) {
-            metrics.get(key).addValue(value);
-        } else {
-            metrics.put(key, new MetricDefinition(key, unit, value));
-        }
+        metrics.compute(
+                key,
+                (k, v) -> {
+                    if (v == null) return new MetricDefinition(key, unit, value);
```

Comment:
does this PR cover the `values` field to be synchronized list in `MetricDefinition`?

Reply:
Oh that is basically because we have make the `metrics` field in `MetricDirective` a ConcurrentHashMap and rewrite `putMetric()` using atomic operation. When adding metrics with the same key, it will first need to access the map with the same key (i.e the same bucket) with an atomic update as we have performed here, which is synchronized by the ConcurrentHashMap itself. As a result, there's no possibility for multiple threads to modify the underlying list in MetricDefinition concurrently anymore.